backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/impl/AbstractLuceneDocumentValueSort.java

```diff
@@ -32,24 +34,27 @@ public abstract class AbstractLuceneDocumentValueSort extends AbstractLuceneReve
 
 	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
 
-	protected final SortField sortField;
+	protected final Function<LuceneSearchSortContext, LuceneFieldComparatorSource> sortFieldCreator;
+	private final String absoluteFieldPath;
 
 	protected AbstractLuceneDocumentValueSort(AbstractBuilder builder) {
 		super( builder );
-		LuceneFieldComparatorSource fieldComparatorSource = builder.toFieldComparatorSource();
-		sortField = new SortField( builder.absoluteFieldPath, fieldComparatorSource, order == SortOrder.DESC );
+		sortFieldCreator = builder::toFieldComparatorSource;
```

Comment:
You just shifted the problem around xD. In things like this:

https://github.com/hibernate/hibernate-search/pull/3964/files#diff-28633312525f6cf27d6995d8708c75770f7a91529e01c66fb19e3b55a2b242beR105-R108

```
		protected Function<SortRequestContext, LuceneFieldComparatorSource> toFieldComparatorSource() {
			return context -> new LuceneGeoPointDistanceComparatorSource( nestedDocumentPath, center,
					getEffectiveMissingValue(),
					getMultiValueMode(), getNestedFilter( context )
```

... the lambda still holds a reference to the builder, so that it'll be able to call `getEffectiveMissingValue()`.

Really, you'll need to either move `toFieldComparatorSource` to the sort itself (passing additional arguments to the constructor), or to call these `getEffectiveMissingValue()` methods outside of the lambda to avoid capturing a reference to the builder. Though the latter seems more prone to future mistakes.